### ISSUE TITLE
postgres.js needs something like slonik's sql.join() for building dynamic queries

### ISSUE THREAD

**Issue by DanielFGray** [porsager/postgres#807]
postgres.js gives us parameterized tagged templates and the ability to nest sql partials, and this is fantastic, but I think it is missing a piece of the puddle:

I ~~want~~ *need* to be able concat partials together (that may contain query parameters), usually with some delimiter (eg whitespace, or a comma, or maybe an sql operator)

slonik provides something like:
```js
sql.join = (a: PendingQuery[] | Helper[], p: PendingQuery): PendingQuery {...}
```
and I think it would be amazing if this were a first-class feature in postgres.js

this should allow things like
```js
let dynamicFilters = [
  sql`somecol = ${somevalue}`,
  sql`col2 = ${v2}`
];
const dynamicColumns = [sql('somecol'), sql('col2')]
await sql`
  select ${sql.join(dynamicColumns, sql`, `)} from t
  where ${sql.join(dynamicFilters, sql` and `)};
`;
```

**Comment by porsager** [comment]
Yeah, I think it makes sense to add, but I've just been doing this when needed

```js
xs.flatMap((x, i) => i ? [sql`, `, x] : x)
```

You can fairly quickly make a function to use instead if you feel like it - eg 😋

```js
sql.join = (xs, joiner) => xs.flatMap((x, i) => i ? [joiner, x] : x)
```

I'll keep the issue open for now.

**Comment by DanielFGray** [comment]
> `sql.join = (xs, joiner) => xs.flatMap((x, i) => i ? [joiner, x] : x)`

oh interesting! i would never have come up with that on my own

do you have any reservations about adding this as a core feature?

**Comment by DanielFGray** [comment]
and if no reservations, would you accept a PR adding this as a core feature?

**Comment by davepar** [comment]
It would be great to at least add the work-around to the docs. It wasn't obvious to me how to join sql fragments with a comma. Glad I found this.

**Comment by DanielFGray** [comment]
i'm working on a PR

anyone want to bike-shed over `join` vs `concat` to avoid confusion with sql joins?

**Comment by Destreyf** [comment]
> Yeah, I think it makes sense to add, but I've just been doing this when needed
> 
> ```js
> xs.flatMap((x, i) => i ? [sql`, `, x] : x)
> ```
> 
> You can fairly quickly make a function to use instead if you feel like it - eg 😋
> 
> ```js
> sql.join = (xs, joiner) => xs.flatMap((x, i) => i ? [joiner, x] : x)
> ```
> 
> I'll keep the issue open for now.

This was insanely helpful, I have a really nested structure so my version is _slightly_ different, and I am monkey patching the sql/database variable.

```ts
// Dumb placeholder to shorten code
export type PendingQuery<T extends postgres.Row = postgres.Row> =
  postgres.PendingQuery<T[]>;

// Hack to add the "join" method.
export type sql = ReturnType<typeof postgres> & {
  join: (a: Array<PendingQuery | null>, b: PendingQuery) => PendingQuery;
};

const db = postgres(config) as sql;

db["join"] = (fragments: (PendingQuery | null)[], joiner: PendingQuery) =>
  fragments.flatMap((x, i) => {
    if (!i) return x;

    if (Array.isArray(x)) {
      // already an array, means we're nested, so lets spread it
      return [joiner, ...x];
    }

    return [joiner, x];
  }) as unknown as PendingQuery;
```

It's typescript, but you can strip the types and it should work, usage is:

```js
db.join([
  db`name = 'test1'`,
  db`name = 'test2'`,
  db`name = 'test3'`,
  db.join([
    db`updated_at >= 2024-01-01`,
    db`updated_at <= '2024-12-31`
  ],db`AND`)
], db`OR`)
```

Hopefully this can make it into somewhere in the codebase as it's super useful for report building.

@DanielFGray hopefully this code can help if you haven't figured out this part.

**Comment by awildeep** [comment]
Any word on making this happen without hacks?

**Comment by FrancescoDiMuro** [comment]
> Any word on making this happen without hacks?
It's crazy annoying to join where clauses.
Any suggestions?
